app/vlselect/logsql/logsql.go

```diff
@@ -542,30 +580,47 @@ func (tp *tailProcessor) getTailRows() ([][]logstorage.Field, error) {
 		return nil, tp.err
 	}
 
-	var resultRows []logRow
+	resultRows := tp.resultRows[:0]
+	emptyStreams := 0
 	for streamID, rows := range tp.perStreamRows {
+		if len(rows) == 0 {
+			emptyStreams++
+			continue
+		}
 		sortLogRows(rows)
 
 		lastTimestamp, ok := tp.lastTimestamps[streamID]
 		if ok {
 			// Skip already written rows
-			for len(rows) > 0 && rows[0].timestamp <= lastTimestamp {
-				rows = rows[1:]
+			i, _ := slices.BinarySearchFunc(rows, lastTimestamp, func(row logRow, timestamp int64) int {
```

Comment:
This optimization is irrelevant for this particular case, since it is expected that live tailing returns logs at low rate.

Also, this optimization complicates code significantly comparing to the original code, so it would be great to include benchmarks in this PR, which clearly show the benefit of this optimization.